libusb/os/linux_android_jni.c

```diff
@@ -0,0 +1,1054 @@
+
+		r =
+			android_jni_gen_config(jni, jni_env, config,
+				descriptors, descriptors_len,
+				strings, strings_len);
+
+		(*jni_env)->DeleteLocalRef(jni_env, config);
+
+		if (r != LIBUSB_SUCCESS)
+			return r;
+	}
+
+	return LIBUSB_SUCCESS;
+}
+
+int android_jni_detect_permission(struct android_jni_context *jni, jobject device,
+	int *has_permission)
+{
+	int r;
+	JNIEnv *jni_env;
+
+	r = android_jni_env(jni, &jni_env);
+	if (r != LIBUSB_SUCCESS)
+		return r;
+
+	/* boolean has_permission = usb_manager.hasPermission(device); */
+	*has_permission =
+		(*jni_env)->CallBooleanMethod(jni_env,
+			jni->usb_manager, jni->UsbManager_hasPermission, device);
+
+	return LIBUSB_SUCCESS;
+}
+
+int android_jni_request_permission(struct android_jni_context *jni,
+	jobject device)
+{
+	int r;
+	JNIEnv *jni_env;
+	jobject intent, permission_intent;
+
+	r = android_jni_env(jni, &jni_env);
+	if (r != LIBUSB_SUCCESS)
+		return r;
+
+	/* Intent intent = new Intent(permission_action); */
+	intent =
+		(*jni_env)->NewObject(jni_env,
+			jni->Intent, jni->Intent_init,
+			jni->permission_action);
+
+	/* PendingIntent permission_intent =
+ 		PendingIntent.getBroadcast(application_context, 0, intent, 0); */
+	permission_intent =
+		(*jni_env)->CallStaticObjectMethod(jni_env,
+			jni->PendingIntent, jni->PendingIntent__getBroadcast,
+			jni->application_context, 0, intent, 0);
+
+	(*jni_env)->DeleteLocalRef(jni_env, intent);
+
+	/* usb_manager.requestPermission(device, permission_intent); */
+	(*jni_env)->CallVoidMethod(jni_env,
+		jni->usb_manager, jni->UsbManager_requestPermission,
+		device, permission_intent);
+
+	(*jni_env)->DeleteLocalRef(jni_env, permission_intent);
+
+	return LIBUSB_SUCCESS;
+}
+
+int android_jni_connect(struct android_jni_context *jni,
+	jobject device, jobject *connection,
+	int *fd, uint8_t **descriptors, size_t *descriptors_len)
+{
+	int has_permission, r;
+	JNIEnv *jni_env;
+	jobject local_connection;
+	jbyteArray local_descriptors;
+	jbyte * local_descriptors_ptr;
+
+	r = android_jni_env(jni, &jni_env);
+	if (r != LIBUSB_SUCCESS)
+		return r;
+
+	r = android_jni_detect_permission(jni, device, &has_permission);
+
+	if (r != LIBUSB_SUCCESS)
+		return r;
+
+	if (!has_permission)
+		return LIBUSB_ERROR_ACCESS;
+
+	/* UsbDeviceConnection local_connection = usb_manager.openDevice(device); */
+	local_connection =
+		(*jni_env)->CallObjectMethod(jni_env,
+			jni->usb_manager, jni->UsbManager_openDevice, device);
+
+	/* fd output */
+	/* fd = local_connection.getFileDescriptor(); */
+	*fd =
+		(*jni_env)->CallIntMethod(jni_env,
+			local_connection, jni->UsbDeviceConnection_getFileDescriptor);
+
+	if (*fd == -1) {
+		(*jni_env)->DeleteLocalRef(jni_env, local_connection);
+		return LIBUSB_ERROR_IO;
+	}
+	
+	/* byte[] local_descriptors = local_connection.getRawDescriptors(); */
+	local_descriptors =
+		(*jni_env)->CallObjectMethod(jni_env,
+			local_connection, jni->UsbDeviceConnection_getRawDescriptors);
```

Comment:
@Youw This is where correct descriptors are downloaded from the device.  The purpose of accessing strings before this was to leave people something to use if e.g. #875 happened.

Reply:
Ah, that's why `parse_config_descriptors` is done twise (once when enumerated, and secondly when actually opened). I've 
been reviewing the implementation closely, and didn't catch this moment.

Hm, so in theory, in my example above, I should be able to `libusb_get_device_descriptor` _after_ I open the device and get the correct result. I'll try.

Reply:
Yeap, that works.
I've event prepared a good public example: https://github.com/libusb/hidapi/pull/273